packager/packager.cc

```diff
@@ -2074,13 +2074,39 @@ void Packager::setPackageNameOnFiles(core::GlobalState &gs, absl::Span<const cor
     }
 }
 
-void Packager::run(core::GlobalState &gs, WorkerPool &workers, absl::Span<ast::ParsedFile> files) {
+namespace {
+
+enum class PackagerMode {
+    PackagesOnly,
+    PackagedFilesOnly,
+    AllFiles,
+};
+
+template <PackagerMode Mode>
+void packageRunCore(core::GlobalState &gs, WorkerPool &workers, absl::Span<ast::ParsedFile> files) {
     ENFORCE(!gs.runningUnderAutogen, "Packager pass does not run in autogen");
 
     Timer timeit(gs.tracer(), "packager");
 
-    findPackages(gs, files);
-    setPackageNameOnFiles(gs, files);
+    switch (Mode) {
+        case PackagerMode::PackagesOnly:
+            timeit.setTag("mode", "packages_only");
+            break;
+        case PackagerMode::PackagedFilesOnly:
+            timeit.setTag("mode", "packaged_files_only");
+            break;
+        case PackagerMode::AllFiles:
+            break;
+    }
+
+    constexpr bool buildPackageDB = Mode == PackagerMode::PackagesOnly || Mode == PackagerMode::AllFiles;
+    constexpr bool validatePackagedFiles = Mode == PackagerMode::PackagedFilesOnly || Mode == PackagerMode::AllFiles;
+
+    if constexpr (buildPackageDB) {
+        Packager::findPackages(gs, files);
+    }
+
+    Packager::setPackageNameOnFiles(gs, files);
```

Comment:
I'd like to only run this when `validatePackagedFiles` is true, but I couldn't determine if we need to have the package name set for `__package.rb` files as well.